Re-work egress firewall to use ACLs + miscellaneous fixes

alexanderConstantinescu opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/ovn-org/ovn-kubernetes/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Trivial changes are exempt from following this template.
If your change is non-trivial, please provide the following information:
-->

**- What this PR does and why is it needed**
<!--
A summary of the changes within this pull request and some context
as to why they were made
-->

This PR performs the following major changes for egress firewall:

- Changes egress firewall in local gateway mode to use ACLs on the nodes' logical switch. As supposed to PR: #1901 which applied that on the join switch: we cannot do that in local gateway mode since traffic does not flow through the join switch. The genesis for this change is because previous implementation of egress firewall as logical router policies conflicted with egress IP and rendered the features mutually exclusive if one pod happened to match both.
- It re-works the way egress firewall filters out east <-> west traffic. Previously this was done by using an inequality operator on the match condition. This is however dangerous as it can lead to an excessive amount of flows. As to avoid doing that I am introducing a set of default allow ACL rules which will never block any east <-> west traffic, no matter the egress firewall specified. This also deserves a unanimous discussion here as previously this was not entirely the case. We never allowed pod -> host network pod traffic when a "deny 0.0.0.0/0" rule was applied. This however seemed counter-intuitive as egress firewall concerns itself with egress traffic only and defines no clear behavior for such traffic. 

After this patch we'll have:

- network policies being applied as ACLs on the logical switch port (and hence evaluated first, giving it highest priority)
- egress network policies being applied as ACLs on the nodes' logical switch 
- egress IP being applied as logical router policies on the the distributed gateway router: ovn_cluster_router (and hence evaluated last if all other *policies allows it)

Besides that there are some minor fixes around this patch with egress IP (which is only in this PR because those functions end up being re-used by the egress firewall code) and various other helpers to make this work. 


/assign @trozet @danwinship @JacobTanenbaum 

**- Special notes for reviewers**
<!--
What exactly did you change - you may also defer to information
contained in commit messages. At a bare minimum it's worth highlighting
which areas of the code were changed as it's easier to assign reviewers
-->

Commit `5bbefc4` and  `26f7f50` can be squashed. But I chose to keep it split during the review as to facilitate the review process. 

**- How to verify it**
<!--
Did you include unit tests? or end-to-end tests?
How can I manually verify that this patch achieves its objective
-->


**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

alexanderConstantinescu commented:
/retest

alexanderConstantinescu commented:
/retest

alexanderConstantinescu commented:
/retest

fedepaol commented on go-controller/pkg/ovn/egressfirewall.go:
```diff
@@ -43,10 +44,10 @@ type destination struct {
 }
 
-func newEgressFirewall(egressFirewallPolicy *egressfirewallapi.EgressFirewall) *egressFirewall {
+func newEgressFirewall(originalEgressfirewall *egressfirewallapi.EgressFirewall) *egressFirewall {
```
can this be `cloneEgressFirewall` or something like that?
I think readability would benefit. Also, the name clashes with `newEgressFirewall` - the variable used in some functions.

Another option is to unwrap the function in the only place it is used

alexanderConstantinescu replied:
I'll unwrap it then, I didn't notice it was only called in one place. Thanks!

alexanderConstantinescu replied:
Actually on second thought, unwrapping it will lead to conflicts between the argument name in the function where it is called and the struct name. I'll just call it `cloneEgressFirewall`

fedepaol commented on test/e2e/e2e_test.go:
```diff
@@ -782,6 +782,204 @@ spec:
+		egressIP := dupIP(egressNodeIP)
+		egressIP[len(egressIP)-2]++
+
+		var egressIPConfig = fmt.Sprintf(`apiVersion: k8s.ovn.org/v1
```
No need to sprintf if you are concatenating the strings

alexanderConstantinescu replied:
You're right, didn't notice it.

fedepaol reviewed (commented):
Left a few comments, hope it helps

JacobTanenbaum commented on go-controller/pkg/ovn/egressfirewall.go:
```diff
@@ -248,80 +249,72 @@ func (oc *Controller) addEgressFirewallRules(hashedAddressSetNameIPv4, hashedAdd
-			fmt.Sprintf("priority=%d", priority),
-			match, "action="+action, fmt.Sprintf("external-ids:egressFirewall=%s", externalID),
-			"--", "add", "logical_router", types.OVNClusterRouter, "policies", "@logical_router_policy")
+		nodes, err := oc.kube.GetNodes()
```
we should use the factories informer cache to get nodes

alexanderConstantinescu replied:
Yes, you're right. Will change

JacobTanenbaum commented on go-controller/pkg/ovn/ovn.go:
```diff
@@ -696,14 +699,33 @@ func (oc *Controller) WatchCRD() {
+	return oc.watchFactory.AddNodeHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			node := obj.(*kapi.Node)
+			oc.addNodeForEgressFirewall(node)
```
I am not sure what the overhead is for handlers is but would it be better to check and add this funtion to add`addNode()` and the corresponding deletion in `deleteNode()`

alexanderConstantinescu replied:
I only did it like this because we don't have a feature flag for egress firewall, and we shouldn't be creating all allow rules on cluster which don't use egress firewall.

alexanderConstantinescu commented:
Something is really strange with our endpoint handlers. I've been looking at the failing CI runs, and I've seen that the test `NetworkPolicy between server and client should enforce policy to allow traffic from pods within server namespace based on PodSelector` has been failing in some builds, here for example: https://github.com/ovn-org/ovn-kubernetes/runs/1761400270?check_suite_focus=true 

All the logs below are taken from ovnkube-master in: https://github.com/ovn-org/ovn-kubernetes/suites/1882356689/artifacts/37064363

Looking at the logs of the run, I am seeing strange messages appear w.r.t our endpoints handlers. 

During the second test execution I am seeing the pod and services created as they should, however the endpoint add event is not called until a long time after (once the test has already failed and the pods and service been removed). This test fails at:

```
Jan 25 11:28:29.470: Pod client-can-connect-80-4slwv should be able to connect to service svc-server, but was not able to connect.
``` 

These are the events we can find in the logs which concern the server pod that the client tries to connect to:

```
grep -ri "server-gpcpx" ovnkube-master.log 
I0125 11:27:37.053896      51 pods.go:254] Creating logical port for network-policy-2362_server-gpcpx on switch ovn-worker2
I0125 11:27:37.054202      51 kube.go:61] Setting annotations map[k8s.ovn.org/pod-networks:{"default":{"ip_addresses":["10.244.2.238/24"],"mac_address":"0a:58:0a:f4:02:ee","gateway_ips":["10.244.2.1"],"ip_address":"10.244.2.238/24","gateway_ip":"10.244.2.1"}}] on pod network-policy-2362/server-gpcpx
I0125 11:27:37.816968      51 port_cache.go:55] port-cache(network-policy-2362_server-gpcpx): added port &{name:network-policy-2362_server-gpcpx uuid:2ae1f777-f5ec-467d-8bf0-d3e9e9a3bdda logicalSwitch:ovn-worker2 ips:[0xc0009c93b0] mac:[10 88 10 244 2 238] expires:{wall:0 ext:0 loc:<nil>}}
I0125 11:27:37.938902      51 pods.go:244] [network-policy-2362/server-gpcpx] addLogicalPort took 885.012432ms
I0125 11:28:38.563881      51 pods.go:78] Deleting pod: network-policy-2362/server-gpcpx
I0125 11:28:38.641181      51 port_cache.go:69] port-cache(network-policy-2362_server-gpcpx): scheduling port for removal at 2021-01-25 11:29:38.641174087 +0000 UTC m=+1264.353580390
I0125 11:29:38.643292      51 port_cache.go:84] port-cache(network-policy-2362_server-gpcpx): removing port
``` 

However, if we look at the endpoint event w.r.t the service:

```
grep -ri "svc-server" ovnkube-master.log  | grep network-policy-2362
I0125 11:27:38.638611      51 service.go:393] Service Reject ACL created for ClusterIP service: svc-server, namespace: network-policy-2362, via: TCP:10.96.226.110:80, ACL UUID: 87843f02-d63f-4c3d-a7b7-0706fbcd6586
I0125 11:27:39.119468      51 service.go:393] Service Reject ACL created for ClusterIP service: svc-server, namespace: network-policy-2362, via: TCP:10.96.226.110:81, ACL UUID: 55ddfb36-3ba4-41e0-89e2-5a6e1f0c0b58
I0125 11:28:58.655172      51 endpoints.go:47] Adding endpoints: svc-server for namespace: network-policy-2362
I0125 11:28:58.655324      51 endpoints.go:54] No service found for endpoint svc-server in namespace network-policy-2362
I0125 11:28:59.095808      51 endpoints.go:47] Adding endpoints: svc-server for namespace: network-policy-2362
I0125 11:28:59.095822      51 endpoints.go:54] No service found for endpoint svc-server in namespace network-policy-2362
I0125 11:29:01.394045      51 endpoints.go:182] Deleting endpoints: svc-server for namespace: network-policy-2362
I0125 11:29:01.394204      51 endpoints.go:188] No service found for endpoint svc-server in namespace network-policy-2362
I0125 11:29:01.394359      51 endpoints.go:182] Deleting endpoints: svc-server for namespace: network-policy-2362
I0125 11:29:01.395727      51 endpoints.go:188] No service found for endpoint svc-server in namespace network-policy-2362

```

This means that the service ADD happens instantaneously, however the endpoint ADD is almost 1.5 min later, which is enough for the test to reach its timeout and fail...

Moreover, if I look in the ovnkube-master logs at the endpoints event around that time ~11:28:58 we see the following:

```
I0125 11:28:58.629977      51 ovs.go:170] exec(9125): stderr: ""
I0125 11:28:58.630134      51 endpoints.go:47] Adding endpoints: svc-server for namespace: network-policy-1036
I0125 11:28:58.630270      51 endpoints.go:54] No service found for endpoint svc-server in namespace network-policy-1036
I0125 11:28:58.630374      51 endpoints.go:182] Deleting endpoints: svc-ns-b-server-b for namespace: network-policy-b-5767
I0125 11:28:58.630523      51 endpoints.go:188] No service found for endpoint svc-ns-b-server-b in namespace network-policy-b-5767
I0125 11:28:58.630663      51 endpoints.go:182] Deleting endpoints: svc-ns-b-server-a for namespace: network-policy-b-5767
I0125 11:28:58.630782      51 endpoints.go:188] No service found for endpoint svc-ns-b-server-a in namespace network-policy-b-5767
I0125 11:28:58.633276      51 endpoints.go:182] Deleting endpoints: svc-ns-b-server-a for namespace: network-policy-b-5767
I0125 11:28:58.638514      51 endpoints.go:188] No service found for endpoint svc-ns-b-server-a in namespace network-policy-b-5767
I0125 11:28:58.638689      51 endpoints.go:182] Deleting endpoints: svc-server for namespace: network-policy-5807
I0125 11:28:58.638847      51 endpoints.go:188] No service found for endpoint svc-server in namespace network-policy-5807
I0125 11:28:58.639518      51 endpoints.go:182] Deleting endpoints: svc-server for namespace: network-policy-5807
I0125 11:28:58.653770      51 endpoints.go:188] No service found for endpoint svc-server in namespace network-policy-5807
I0125 11:28:58.653982      51 endpoints.go:182] Deleting endpoints: svc-server for namespace: network-policy-1036
I0125 11:28:58.654150      51 endpoints.go:188] No service found for endpoint svc-server in namespace network-policy-1036
I0125 11:28:58.654281      51 endpoints.go:47] Adding endpoints: agnhost-primary for namespace: kubectl-8735
I0125 11:28:58.654423      51 endpoints.go:62] Matching service agnhost-primary found for ep: agnhost-primary, with cluster IP: 10.96.233.28
I0125 11:28:58.654547      51 endpoints.go:41] Endpoint Protocol Map is: map[SCTP:map[] TCP:map[] UDP:map[]]
I0125 11:28:58.654707      51 endpoints.go:65] Matching service agnhost-primary ports: [{ TCP <nil> 6379 {1 0 agnhost-server} 0}]
I0125 11:28:58.654881      51 endpoints.go:182] Deleting endpoints: svc-server for namespace: network-policy-1036
I0125 11:28:58.655017      51 endpoints.go:188] No service found for endpoint svc-server in namespace network-policy-1036
I0125 11:28:58.655172      51 endpoints.go:47] Adding endpoints: svc-server for namespace: network-policy-2362
I0125 11:28:58.655324      51 endpoints.go:54] No service found for endpoint svc-server in namespace network-policy-2362
I0125 11:28:58.655460      51 endpoints.go:47] Adding endpoints: agnhost-primary for namespace: kubectl-8735
I0125 11:28:58.655588      51 endpoints.go:62] Matching service agnhost-primary found for ep: agnhost-primary, with cluster IP: 10.96.233.28
I0125 11:28:58.655722      51 endpoints.go:41] Endpoint Protocol Map is: map[SCTP:map[] TCP:map[:{[10.244.2.235] 6379}] UDP:map[]]
I0125 11:28:58.655835      51 endpoints.go:65] Matching service agnhost-primary ports: [{ TCP <nil> 6379 {1 0 agnhost-server} 0}]
I0125 11:28:58.656067      51 loadbalancer.go:113] Creating lb with 02f6435e-9435-46d8-b7ba-45d99bbdb4bb, [[10.96.233.28]], 6379, [[10.244.2.235]], 6379
I0125 11:28:58.656277      51 ovs.go:166] exec(9126): /usr/bin/ovn-nbctl --timeout=15 set load_balancer 02f6435e-9435-46d8-b7ba-45d99bbdb4bb vips:"10.96.233.28:6379"="10.244.2.235:6379"
I0125 11:28:58.753167      51 ovs.go:169] exec(9126): stdout: ""
I0125 11:28:58.753360      51 ovs.go:170] exec(9126): stderr: ""
I0125 11:28:58.753419      51 loadbalancer.go:101] LB entry set for 02f6435e-9435-46d8-b7ba-45d99bbdb4bb, vips:"10.96.233.28:6379"="10.244.2.235:6379", &{[10.244.2.235:6379] b1d56d79-7926-48c9-8f67-4f92b4a6f4cf}
I0125 11:28:58.753531      51 ovs.go:166] exec(9127): /usr/bin/ovn-nbctl --timeout=15 --data=bare --no-heading --columns=name find logical_router load_balancer{>=}02f6435e-9435-46d8-b7ba-45d99bbdb4bb
I0125 11:28:58.759175      51 namespace.go:278] Deleting namespace: network-policy-1036
I0125 11:28:58.763384      51 address_set.go:453] destroy(19e73518-58eb-4f5b-98ae-d3622258beff/network-policy-1036_v4/a7800512977357189504)
I0125 11:28:58.763561      51 ovs.go:166] exec(9128): /usr/bin/ovn-nbctl --timeout=15 --if-exists destroy address_set 19e73518-58eb-4f5b-98ae-d3622258beff
I0125 11:28:58.843092      51 ovs.go:169] exec(9127): stdout: ""
I0125 11:28:58.843276      51 ovs.go:170] exec(9127): stderr: ""
I0125 11:28:58.843593      51 ovs.go:166] exec(9129): /usr/bin/ovn-nbctl --timeout=15 -- --if-exists remove logical_switch  acl b1d56d79-7926-48c9-8f67-4f92b4a6f4cf
I0125 11:28:58.849363      51 ovs.go:169] exec(9128): stdout: ""
I0125 11:28:58.852896      51 ovs.go:170] exec(9128): stderr: ""
I0125 11:28:58.853165      51 namespace.go:286] Deleting Namespace's NetworkPolicy entities
I0125 11:28:58.854001      51 namespace.go:213] Updating namespace: webhook-3859-markers
I0125 11:28:58.899401      51 ovs.go:169] exec(9129): stdout: ""
I0125 11:28:58.899657      51 ovs.go:170] exec(9129): stderr: ""
I0125 11:28:58.899814      51 loadbalancer.go:355] ACL: b1d56d79-7926-48c9-8f67-4f92b4a6f4cf, removed from switches: []
I0125 11:28:58.915216      51 service.go:466] Deleting service latency-svc-chg98
I0125 11:28:58.915528      51 ovs.go:166] exec(9131): /usr/bin/ovn-nbctl --timeout=15 --if-exists remove load_balancer 02f6435e-9435-46d8-b7ba-45d99bbdb4bb vips "10.96.80.16:80"
I0125 11:28:58.915455      51 ovs.go:166] exec(9130): /usr/bin/ovn-nbctl --timeout=15 -- --if-exists remove port_group ce989bcb-0ff7-4bb0-a07a-1feea4bb7e3c acls b1d56d79-7926-48c9-8f67-4f92b4a6f4cf
I0125 11:28:59.083641      51 ovs.go:169] exec(9131): stdout: ""
I0125 11:28:59.083968      51 ovs.go:170] exec(9131): stderr: ""
I0125 11:28:59.084114      51 ovs.go:166] exec(9132): /usr/bin/ovn-nbctl --timeout=15 --data=bare --no-heading --columns=name find logical_router load_balancer{>=}02f6435e-9435-46d8-b7ba-45d99bbdb4bb
I0125 11:28:59.095643      51 ovs.go:169] exec(9130): stdout: ""
I0125 11:28:59.095683      51 ovs.go:170] exec(9130): stderr: ""
I0125 11:28:59.095698      51 loadbalancer.go:365] ACL: b1d56d79-7926-48c9-8f67-4f92b4a6f4cf, removed from the port group : ce989bcb-0ff7-4bb0-a07a-1feea4bb7e3c
I0125 11:28:59.095735      51 endpoints.go:47] Adding endpoints: service-proxy-disabled for namespace: services-2504
I0125 11:28:59.095758      51 endpoints.go:54] No service found for endpoint service-proxy-disabled in namespace services-2504
I0125 11:28:59.095768      51 endpoints.go:182] Deleting endpoints: svc-server for namespace: network-policy-2494
I0125 11:28:59.095780      51 endpoints.go:188] No service found for endpoint svc-server in namespace network-policy-2494
I0125 11:28:59.095788      51 endpoints.go:182] Deleting endpoints: svc-server for namespace: network-policy-2494
I0125 11:28:59.095798      51 endpoints.go:188] No service found for endpoint svc-server in namespace network-policy-2494
I0125 11:28:59.095808      51 endpoints.go:47] Adding endpoints: svc-server for namespace: network-policy-2362
I0125 11:28:59.095822      51 endpoints.go:54] No service found for endpoint svc-server in namespace network-policy-2362
I0125 11:28:59.095844      51 endpoints.go:47] Adding endpoints: service-proxy-disabled for namespace: services-2504
I0125 11:28:59.095858      51 endpoints.go:54] No service found for endpoint service-proxy-disabled in namespace services-2504
I0125 11:28:59.095870      51 endpoints.go:47] Adding endpoints: service-proxy-disabled for namespace: services-2504
I0125 11:28:59.095882      51 endpoints.go:54] No service found for endpoint service-proxy-disabled in namespace services-2504
I0125 11:28:59.095890      51 endpoints.go:47] Adding endpoints: service-proxy-disabled for namespace: services-2504
I0125 11:28:59.095906      51 endpoints.go:54] No service found for endpoint service-proxy-disabled in namespace services-2504
I0125 11:28:59.095915      51 endpoints.go:47] Adding endpoints: service-proxy-toggled for namespace: services-2504
I0125 11:28:59.095924      51 endpoints.go:62] Matching service service-proxy-toggled found for ep: service-proxy-toggled, with cluster IP: 10.96.75.143
I0125 11:28:59.095935      51 endpoints.go:41] Endpoint Protocol Map is: map[SCTP:map[] TCP:map[] UDP:map[]]
I0125 11:28:59.095952      51 endpoints.go:65] Matching service service-proxy-toggled ports: [{ TCP <nil> 80 {0 9376 } 0}]
I0125 11:28:59.095972      51 endpoints.go:47] Adding endpoints: e2e-test-webhook for namespace: webhook-3859
I0125 11:28:59.095981      51 endpoints.go:62] Matching service e2e-test-webhook found for ep: e2e-test-webhook, with cluster IP: 10.96.90.150
I0125 11:28:59.095989      51 endpoints.go:41] Endpoint Protocol Map is: map[SCTP:map[] TCP:map[:{[10.244.2.242] 8444}] UDP:map[]]
I0125 11:28:59.096003      51 endpoints.go:65] Matching service e2e-test-webhook ports: [{ TCP <nil> 8443 {0 8444 } 0}]
I0125 11:28:59.096015      51 loadbalancer.go:113] Creating lb with 02f6435e-9435-46d8-b7ba-45d99bbdb4bb, [[10.96.90.150]], 8443, [[10.244.2.242]], 844
...
...
I0125 11:29:01.394045      51 endpoints.go:182] Deleting endpoints: svc-server for namespace: network-policy-2362
I0125 11:29:01.394204      51 endpoints.go:188] No service found for endpoint svc-server in namespace network-policy-2362
I0125 11:29:01.394359      51 endpoints.go:182] Deleting endpoints: svc-server for namespace: network-policy-2362
I0125 11:29:01.395727      51 endpoints.go:188] No service found for endpoint svc-server in namespace network-policy-2362
I0125 11:29:01.395889      51 endpoints.go:47] Adding endpoints: svc-server for namespace: network-policy-3807
I0125 11:29:01.396040      51 endpoints.go:62] Matching service svc-server found for ep: svc-server, with cluster IP: 10.96.176.240
I0125 11:29:01.396174      51 endpoints.go:41] Endpoint Protocol Map is: map[SCTP:map[] TCP:map[] UDP:map[]]
I0125 11:29:01.396316      51 endpoints.go:65] Matching service svc-server ports: [{server-80 TCP <nil> 80 {0 80 } 0} {server-81 TCP <nil> 81 {0 81 } 0}]
I0125 11:29:01.396463      51 endpoints.go:47] Adding endpoints: nodeport-update-service for namespace: services-7119
I0125 11:29:01.396593      51 endpoints.go:62] Matching service nodeport-update-service found for ep: nodeport-update-service, with cluster IP: 10.96.15.231
I0125 11:29:01.396718      51 endpoints.go:41] Endpoint Protocol Map is: map[SCTP:map[] TCP:map[] UDP:map[]]
I0125 11:29:01.396850      51 endpoints.go:65] Matching service nodeport-update-service ports: [{tcp-port TCP <nil> 80 {0 9376 } 31526}]
I0125 11:29:01.396991      51 endpoints.go:47] Adding endpoints: svc-server for namespace: network-policy-3807
I0125 11:29:01.397126      51 endpoints.go:62] Matching service svc-server found for ep: svc-server, with cluster IP: 10.96.176.240
I0125 11:29:01.397270      51 endpoints.go:41] Endpoint Protocol Map is: map[SCTP:map[] TCP:map[] UDP:map[]]
I0125 11:29:01.398060      51 endpoints.go:65] Matching service svc-server ports: [{server-80 TCP <nil> 80 {0 80 } 0} {server-81 TCP <nil> 81 {0 81 } 0}]
I0125 11:29:01.398244      51 endpoints.go:47] Adding endpoints: svc-server for namespace: network-policy-3807
I0125 11:29:01.398397      51 endpoints.go:62] Matching service svc-server found for ep: svc-server, with cluster IP: 10.96.176.240
I0125 11:29:01.398529      51 endpoints.go:41] Endpoint Protocol Map is: map[SCTP:map[] TCP:map[server-80:{[10.244.0.232] 80} server-81:{[10.244.0.232] 81}] UDP:map[]]
I0125 11:29:01.398677      51 endpoints.go:65] Matching service svc-server ports: [{server-80 TCP <nil> 80 {0 80 } 0} {server-81 TCP <nil> 81 {0 81 } 0}]
I0125 11:29:01.398896      51 loadbalancer.go:113] Creating lb with 02f6435e-9435-46d8-b7ba-45d99bbdb4bb, [[10.96.176.240]], 80, [[10.244.0.232]], 80

```

meaning our endpoint handler events seem to be "bulk added", and multiple times over (i.e we get the ADD event for the endpoint twice, but much later than the service ADD). What I mean by that is that this behavior (i.e: the endpoints event lagging after the service events) happens for multiple services. 

I am not sure if this is common or even known? @dave-tucker ? 

This has nothing to do with my patch though.

dave-tucker commented:
@alexanderConstantinescu there are a handful of known issues with event handling right now that interact with each other in weird and interesting ways. In short:

1. We process every event from k8s for each item - so Add/Update/Update/Update/Update/Update for an item is 1 add and 5 updates. Updates can happen due to many different reasons - sometimes nothing has actually changed.
This creates a rather large tail of events.

2. Our updates are not idempotent and in most cases we instead call add/delete. In the case above, that would mean that for 6 events, we actually call 6 x adds and 5 x deletes.

3. Our event processing code is slow because:
a. we have to wait for external systems (iptables, ovn, ovs etc...) to respond
b. we have some dependencies between updates that can cause us to idle waiting for another event to be processed first

4. Only pod and node events are serviced by > 1 go routine.

5. ResourceEventHandlerFuncs that are added after the informers are started must complete their `SyncX` methods before they start processing events. Depending on what's in that sync method, some events could be missed.

I'm trying to address 1 (and possibly 5) by introducing workqueues. This also adds instrumentation to address 4.

alexanderConstantinescu commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -889,13 +883,14 @@ func (e *egressIPDial) dial(ip net.IP) bool {
-func getClusterSubnets() (*net.IPNet, *net.IPNet) {
-	var v4ClusterSubnet, v6ClusterSubnet *net.IPNet
+func getClusterSubnets() ([]*net.IPNet, []*net.IPNet) {
+	var v4ClusterSubnet = []*net.IPNet{}
```
It can

alexanderConstantinescu commented on go-controller/pkg/ovn/egressfirewall.go:
```diff
@@ -248,80 +249,72 @@ func (oc *Controller) addEgressFirewallRules(hashedAddressSetNameIPv4, hashedAdd
 		if err != nil {
-			return fmt.Errorf("error deleting egressFirewall with external-ids %s, cannot get logical router policies from LR %s - %s:%s",
-				externalID, types.OVNClusterRouter, err, stderr)
+			deletionErrors = errors.Wrapf(deletionErrors, "unable to setup egress firewall ACLs on cluster nodes, err: %v", err)
```
I am not sure, I tried to keep it in line with what has previously been done in `egressfirewall.go`. I'll keep it like this for now

fedepaol replied:
Yep, but in the previous version it was returning. 
That's why I made the comment

alexanderConstantinescu commented on go-controller/pkg/ovn/egressfirewall.go:
```diff
@@ -332,6 +325,180 @@ func deleteEgressFirewallRules(externalID string) error {
+	}
+}
+
+// deleteNodeForEgressFirewall deletes the default allow ACLs as follows:
```
Yes, I tried the second option you mentioned when implementing this, but I felt it obfuscated the method more than increasing the readability. Given that the arguments to `RunOVNNbctl` are not the same if we perform an add/delete, I think I'll keep it like this for now since I don't think splitting the function in two provides a lot of benefit for such a short function.

alexanderConstantinescu commented:
Thanks for explaining @dave-tucker

With the help from @trozet we were able to track down the problem to "infra flakes" yesterday. It seems the API server was also generating the endpoints events with a big latency. We don't know what the cause of that was (most likely github infrastructure being slow), but it didn't point to any problems with our informers.

trozet commented on go-controller/pkg/ovn/egressfirewall.go:
```diff
@@ -43,10 +43,10 @@ type destination struct {
 }
 
-func newEgressFirewall(egressFirewallPolicy *egressfirewallapi.EgressFirewall) *egressFirewall {
+func cloneEgressFirewall(originalEgressfirewall *egressfirewallapi.EgressFirewall) *egressFirewall {
```
doctext

trozet commented on test/e2e/e2e_test.go:
```diff
@@ -782,6 +782,204 @@ spec:
+            name: ` + f.Namespace.Name + `
+`
+
+		if err := ioutil.WriteFile(egressIPYaml, []byte(egressIPConfig), 0644); err != nil {
```
can we do this without writing to disk?

alexanderConstantinescu replied:
As I see it: not without some weird go-bash-combined hackery. `kubectl` does allow you to create an object from stanza which is written to the process STDIN file descriptor. But doing that from here is severe overkill in my opinion. The files are cleaned once the tests are done anyways.

trozet commented on go-controller/pkg/types/const.go:
```diff
@@ -32,6 +32,10 @@ const (
 	NodeLocalSwitch = "node_local_switch"
 
+	// ACL directions
+	DirectionToLPort   = "to-lport"
```
I don't really expect people to migrate to using this constant. I would rather seem them migrate to go-ovn bindings.

alexanderConstantinescu replied:
Not sure what you are referring to? This line is untouched by this commit

trozet replied:
The commit shows you added this line:
```
	DirectionToLPort   = "to-lport"
```

alexanderConstantinescu replied:
Ah, sorry. Must have been my github review layout which confused me. 

> I would rather seem them migrate to go-ovn bindings.

So you want me to remove this commit? The reason why I introduced it is to avoid having cyclic dependencies between the packages consuming this constant.

trozet replied:
no, you dont have to remove it. I was just stating that it will probably go away when we move to bindings.

trozet commented on go-controller/pkg/ovn/egressfirewall.go:
```diff
@@ -251,11 +252,11 @@ func (oc *Controller) addEgressFirewallRules(hashedAddressSetNameIPv4, hashedAdd
 	logicalSwitches := []string{}
 	if config.Gateway.Mode == config.GatewayModeLocal {
-		nodes, err := oc.kube.GetNodes()
+		nodes, err := oc.watchFactory.GetNodes()
```
this should be part of your first commit

alexanderConstantinescu replied:
ACK, I will squash

trozet commented on go-controller/pkg/ovn/egressfirewall.go:
```diff
@@ -324,6 +325,180 @@ func (oc *Controller) deleteEgressFirewallRules(externalID string) error {
+	matchConditions := []string{}
+	if v4ServiceSubnet != nil {
+		for _, v4Subnet := range v4ClusterSubnets {
+			matchConditions = append(matchConditions, fmt.Sprintf("ip4.src == %s && ip4.dst == %s", v4Subnet.String(), v4ServiceSubnet.String()))
```
in these functions, why do you care about matching source? Shouldn't just dest be enough?

alexanderConstantinescu replied:
Because I want to make sure it only matches packets for which the source IP is a pod IP.

trozet replied:
What other traffic would be there? mp0 would be snat'ed to the pod subnet anyway, so that one is out. Is it to avoid matching on from external to pod service/subnet as a response? What about incoming traffic from outside the cluster that would be snat'ed to join switch then forwarded into the switch towards a pod? I need to think about these some more...

alexanderConstantinescu replied:
Hmm, good question, I overlooked the fact that I am doing this on the GR. Let me run some tests and double check the outcome of removing the source IP from the filter.

alexanderConstantinescu replied:
We can remove the source IP. My tests show that it's fine, and logically given what has been stated here it should also be fine.

trozet replied:
@alexanderConstantinescu meant ovn worker switch

I'm wondering if we need allow-related on these ACLs. Let's take an example where a pod is an endpoint to a service.
external traffic -> GR LB -> pod, and lets say there is an egress firewall on the pod

The allow will allow the traffic to reach the pod, but then the pod response will be pod -> external IP and will get dropped. Shouldn't it have allow-related to allow it to send the response?

@dceara do you mind reviewing this. ACLs are tricky and can have some implications. I'd appreciate your input here.

trozet replied:
@alexanderConstantinescu what is the resolution here? are we no longer optimizing ACLs? Removing inverse matching? Maybe we need to have a sync up meeting on this.

alexanderConstantinescu replied:
> are we no longer optimizing ACLs? 

Yes, I removed all the commits related to that.

trozet commented on go-controller/pkg/ovn/egressfirewall.go:
```diff
@@ -324,6 +325,180 @@ func (oc *Controller) deleteEgressFirewallRules(externalID string) error {
+	return matchConditions
+}
+
+func (oc *Controller) generateDefaultNodeAllowMatch(addNode *v1.Node, v4ClusterSubnets, v6ClusterSubnets []*net.IPNet) []string {
```
doctext on this function is missing and I'm going to interpret this as you are trying to add an allow ACL for packets destined to a single node ip? Then the caller is either applying that to the join switch (shared gw) or node switches?

alexanderConstantinescu replied:
Yes, correct. Concerning the doctext, I wrote a doctext on all functions calling this one: `addNodeForEgressFirewall` / `deleteNodeForEgressFirewall` / `deleteDefaultAllowACLsForEgressFirewall`, do you still feel that I should add a doctext here?

trozet replied:
please

trozet reviewed (changes requested):
I have a more general question about: Pod -> host networked pod traffic

Is it explicitly stated somewhere in OpenShift docs that this is the required behavior? From my understanding of this PR, we are allowing pods to bypass egress firewall in order to talk to node IPs, but not other things on the same external subnet. This seems a little odd to me. I think if we can simplify this to allow pods to talk to the node subnet, then we dont need a watcher for node IPs and we can just add a single ACL to cover the node subnet.

alexanderConstantinescu commented:
> I have a more general question about: Pod -> host networked pod traffic
> 
> Is it explicitly stated somewhere in OpenShift docs that this is the required behavior? From my understanding of this PR, we are allowing pods to bypass egress firewall in order to talk to node IPs, but not other things on the same external subnet. This seems a little odd to me. I think if we can simplify this to allow pods to talk to the node subnet, then we dont need a watcher for node IPs and we can just add a single ACL to cover the node subnet.

I am not sure what you're referring to when you mention the "node subnet", in the case of egress firewall and egress IP we don't care about "the IP range allocated to pods on a given node" (which is usually qualified as "node subnet"). We care about the IP range for which nodes get their attributed IP address, and this is something we don't know in ovn-kubernetes (at least to the best of my knowledge). Do you have an example of where this information is used in our code?

alexanderConstantinescu commented on go-controller/pkg/ovn/egressfirewall.go:
```diff
@@ -248,80 +248,72 @@ func (oc *Controller) addEgressFirewallRules(hashedAddressSetNameIPv4, hashedAdd
+	} else {
+		logicalSwitches = append(logicalSwitches, types.OVNJoinSwitch)
+	}
+	uuid, stderr, err := util.RunOVNNbctl("--data=bare", "--no-heading",
```
Yes, and this is done on L271.

trozet replied:
this check isn't good enough for error. Find does not return an error if it doesnt find anything. The value of uuid will just be empty.

alexanderConstantinescu commented on go-controller/pkg/ovn/egressfirewall.go:
```diff
@@ -396,12 +396,11 @@ func generateMatch(ipv4Source, ipv6Source string, destinations []matchTarget, ds
 		match = fmt.Sprintf("%s && %s", match, egressGetL4Match(dstPorts))
 	}
 
-	if config.Gateway.Mode == config.GatewayModeLocal {
```
Yes, I am squashing now and will push an update.

alexanderConstantinescu commented:
/retest

trozet commented:
> his in OVN-Kubernetes today (at least to the best of my knowledge). Do you have an example of where this information is used in our code?

I didn't mean node subnet as in a small range of IPs just used for our nodes. I meant allowing external traffic to all IPs on the external subnet. So you would still block egress traffic to the internet and other traffic routed out, but you would be able to access anything on the directly connected subnet to the nodes. Is that acceptable or is that even too loose? Diagram might make it easier:

```
---internet-----GW 10.0.0.1/24------- 10.0.0.2/24 OVN Node
                                  |
        Server A 10.0.0.200/24 ------ 10.0.0.3/24 OVN Node 2
```
In this scenario pod would be able to reach any of the 10.0.0.x/24 addresses, but it would not be able to go to any other subnet (other than pod/svc subnet). Is  this OK or too loose because the pod can access something like Server A on the external network?

alexanderConstantinescu commented:
> > his in OVN-Kubernetes today (at least to the best of my knowledge). Do you have an example of where this information is used in our code?
> 
> I didn't mean node subnet as in a small range of IPs just used for our nodes. I meant allowing external traffic to all IPs on the external subnet. So you would still block egress traffic to the internet and other traffic routed out, but you would be able to access anything on the directly connected subnet to the nodes. Is that acceptable or is that even too loose? Diagram might make it easier:
> 
> ```
> ---internet-----GW 10.0.0.1/24------- 10.0.0.2/24 OVN Node
>                                   |
>         Server A 10.0.0.200/24 ------ 10.0.0.3/24 OVN Node 2
> ```
> 
> In this scenario pod would be able to reach any of the 10.0.0.x/24 addresses, but it would not be able to go to any other subnet (other than pod/svc subnet). Is this OK or too loose because the pod can access something like Server A on the external network?

I am not sure I follow your example. Server A should be able to talk to `OVN Node` and `OVN Node 2` and any other known node in the cluster (this is all internal cluster traffic). Any IP on the internet or node on the same subnet as the machine CIDR (but not a known OpenShift node) needs to specifically be allowed by the egress firewall if there's a `0.0.0.0/0 block all rule` defined.

Does that answer your question?

alexanderConstantinescu commented:
/retest

trozet commented:
> > > his in OVN-Kubernetes today (at least to the best of my knowledge). Do you have an example of where this information is used in our code?
> > 
> > 
> > I didn't mean node subnet as in a small range of IPs just used for our nodes. I meant allowing external traffic to all IPs on the external subnet. So you would still block egress traffic to the internet and other traffic routed out, but you would be able to access anything on the directly connected subnet to the nodes. Is that acceptable or is that even too loose? Diagram might make it easier:
> > ```
> > ---internet-----GW 10.0.0.1/24------- 10.0.0.2/24 OVN Node
> >                                   |
> >         Server A 10.0.0.200/24 ------ 10.0.0.3/24 OVN Node 2
> > ```
> > 
> > 
> > In this scenario pod would be able to reach any of the 10.0.0.x/24 addresses, but it would not be able to go to any other subnet (other than pod/svc subnet). Is this OK or too loose because the pod can access something like Server A on the external network?
> 
> I am not sure I follow your example. `Server A` should be able to talk to `OVN Node` and `OVN Node 2` and any other known node in the cluster (this is all internal cluster traffic). Any IP on the internet or node on the same subnet as the machine CIDR (but not a known cluster node) needs to specifically be allowed by the egress firewall if there's a `0.0.0.0/0 block all rule` defined.
> 
> Does that answer your question?

we resolved this over slack. Looks like the current implementation is our best option.

trozet commented on go-controller/pkg/types/const.go:
```diff
@@ -39,26 +39,33 @@ const (
+	DefaultRoutedMcastAllowPriority = "10013"
 	// Default multicast allow acl rule priority
-	DefaultMcastAllowPriority = "1012"
+	DefaultMcastAllowPriority = "10012"
```
upgrade implications here

alexanderConstantinescu replied:
Good catch, do we have any plans to add an upgrade KIND job? From master HEAD -> branch HEAD for example?

alexanderConstantinescu replied:
Actually come to think of it, it's even worse than only this commit. Whenever anybody (OCP) will upgrade from local gateway mode to shared gateway mode, the entire egress firewall setup will most likely break everything since in local gateway mode everything is implemented using routing rules, whereas in shared gateway mode it's ACL on the join switch...and egress firewall does not sync any of that.

alexanderConstantinescu replied:
I am proposing (though it hurts given the effort its taken me to implement and constantly rebase) that we go for option 1.  OCP will move to shared gateway mode very soon, and as with any software version improvement, I think we can live with 4.7 having some rules which are sub-optimal, which when we move to shared gateway mode will be solved.  

Having this intermediary optimal local gateway mode will clutter the code with an over complicated sync function which will need to sync old local GW mode -> better local GW mode **and** better local GW mode -> shared GW mode. That's not worth the win we get from improving the rules.

I will write up a patch for syncing egress firewall setup from local to shared gateway mode.

trozet replied:
honestly I'm struggling a little to follow, but this seems reasonable to me. Just concerned about corner cases, leftover config, but if you are confident your newly added sync covers it then we should be good.

trozet replied:
Is "Remove inefficient inequality operator in firewall match" already done in shared gateway egress firewall?

alexanderConstantinescu replied:
> Is "Remove inefficient inequality operator in firewall match" already done in shared gateway egress firewall?

It never had it. Only local gateway mode added match conditions where it filtered out traffic using the inequality operator. 

> honestly I'm struggling a little to follow

We can have a discussion concerning this on slack.

danwinship commented:
Not sure what the current state of the code is, but just commenting on some previous comments

> * As to avoid doing that I am introducing a set of default allow ACL rules which will never block any east <-> west traffic, no matter the egress firewall specified. This also deserves a unanimous discussion here as previously this was not entirely the case. We never allowed pod -> host network pod traffic when a "deny 0.0.0.0/0" rule was applied. This however seemed counter-intuitive as egress firewall concerns itself with egress traffic only and defines no clear behavior for such traffic.

EgressNetworkPolicy in openshift-sdn was supposed to document that you couldn't use it to block access to node IPs, although it looks like it doesn't currently say that :frowning_face:. In particular though, it generally doesn't really matter whether you can use EgressNetworkPolicy/EgressFirewall to block the `eth0` IP of a node, since users would still be able to connect to the node's `tun0`/`ovn-k8s-mp0` IP...

> ```
> ---internet-----GW 10.0.0.1/24------- 10.0.0.2/24 OVN Node
>                                   |
>         Server A 10.0.0.200/24 ------ 10.0.0.3/24 OVN Node 2
> ```
> 
> In this scenario pod would be able to reach any of the 10.0.0.x/24 addresses, but it would not be able to go to any other subnet (other than pod/svc subnet). Is this OK or too loose because the pod can access something like Server A on the external network?

That's not OK; this is explicitly a use case for egress firewall; "I have some database/whatever on my intranet that a few pods need access to, but most other pods should be blocked from accessing."

alexanderConstantinescu commented:
> In particular though, it generally doesn't really matter whether you can use EgressNetworkPolicy/EgressFirewall to block the eth0 IP of a node, since users would still be able to connect to the node's tun0/ovn-k8s-mp0 IP...

OK, in that case what I will do (for the ease of the upgrade process and switch from local -> shared gateway mode), is: remove the "default allow to node" rules I am setting up as to allow egress firewall matching pods access to nodes/host networked pod if a "deny all 0.0.0.0/0" rule is created. Currently with such an egress firewall rule: those pods cannot connect to the Kubernetes API service for example, but from my understanding we can live without that (at least until a someone yells at us)

alexanderConstantinescu commented:
Update:

* I've done what I've described above, meaning: I've removed the rule optimization since that would have induced a bunch of sync complexities across OVN-Kube for what concerns the re-evaluated ACL priorities. This thus also means that egress firewall matching pod cannot connect to host network pods, as it's always been. 
* I've added a sync function syncing the migrations needed. 

The plan going forward will be: once we remove local gateway mode I will fix the fact that egress firewall matching pods can't connect to host networked pods

@trozet : PTAL

danwinship commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -891,16 +885,17 @@ func (e *egressIPDial) dial(ip net.IP) bool {
 
-func getClusterSubnets() (*net.IPNet, *net.IPNet) {
-	var v4ClusterSubnet, v6ClusterSubnet *net.IPNet
+func getClusterSubnets() ([]*net.IPNet, []*net.IPNet) {
```
This function is kind of superfluous now; you could just have `createDefaultNoReroutePodPolicies` etc iterate `config.Default.ClusterSubnets` directly, and call `utilnet.IsIPv6CIDR` there in that loop

alexanderConstantinescu replied:
`getClusterSubnets` is also called in `addNodeForEgress` and `deleteNodeForEgress` though

danwinship commented on go-controller/pkg/types/const.go:
```diff
@@ -36,6 +36,19 @@ const (
+	// Default allow acl rule priority
+	DefaultAllowPriority = "1001"
+	// Default deny acl rule priority
+	DefaultDenyPriority = "1000"
```
Consider better naming; none of these are really "Default"s since they can't be overridden. Also, the Allow/Deny priorities are specifically for Allow/Deny rules for NetworkPolicy, not all Allows/Denies.

alexanderConstantinescu replied:
I only moved these to another file, and though I agree that they could use a better naming convention I would like to avoid doing it in this PR as to not delay it further.

danwinship commented on go-controller/pkg/ovn/egressfirewall.go:
```diff
@@ -76,6 +76,92 @@ func newEgressFirewallRule(rawEgressFirewallRule egressfirewallapi.EgressFirewal
 	return efr, nil
 }
 
+/* This function is used to sync egress firewall setup. It does two "cleanups"
```
(use `//` style comments?)

danwinship reviewed (commented):
I did a quick review but I didn't look at any of the actual ACL stuff

trozet commented on go-controller/pkg/ovn/egressfirewall.go:
```diff
@@ -76,6 +76,90 @@ func newEgressFirewallRule(rawEgressFirewallRule egressfirewallapi.EgressFirewal
+//  	For this it just deletes all ACL setup done for egress firewall on the node switches
+
+// NOTE: Utilize the fact that we know that all egress firewall related setup must have a priority: types.MinimumReservedEgressFirewallPriority <= priority <= types.EgressFirewallStartPriority
+func (oc *Controller) syncEgressFirewall(egressFirwalls []interface{}) {
```
perhaps a follow up you may need to check if its possible to upgrade from "old implementation local gw" -> "shared gw" and see if this would handle that scenario.

trozet replied:
thanks

alexanderConstantinescu commented:
/retest

trozet reviewed (approved):
/lgtm me on green

trozet commented:
/retest
